spring-pulsar/src/main/java/org/springframework/pulsar/core/DefaultTopicResolver.java

```diff
@@ -100,9 +116,31 @@ protected Resolved<String> doResolveTopic(@Nullable String userSpecifiedTopic, @
 		if (messageType == null) {
 			return Resolved.failed("Topic must be specified when the message is null");
 		}
-		String topic = this.customTopicMappings.getOrDefault(messageType, defaultTopicSupplier.get());
+		// Check for custom topic mapping
+		String topic = this.customTopicMappings.get(messageType);
+
+		// If no custom topic mapping found, look for @PulsarTypeMapping (if enabled)
+		if (this.usePulsarTypeMappingAnnotations && topic == null) {
+			topic = getAnnotatedTopicInfo(messageType);
+			if (topic != null) {
+				this.addCustomTopicMapping(messageType, topic);
+			}
+		}
+
+		// If still no topic, consult the default topic supplier
+		if (topic == null) {
+			topic = defaultTopicSupplier.get();
+		}
 		return topic == null ? Resolved.failed("Topic must be specified when no default topic is configured")
 				: Resolved.of(topic);
 	}
 
+	// VisibleForTesting
+	String getAnnotatedTopicInfo(Class<?> messageType) {
```

Comment:
Yep, I had it like that originally but mostly because the flow continues at line 130 and changing to optional at that point would ripple the surrounding code (line 120, line 131).